lib/fs.js

```diff
@@ -1097,7 +1097,9 @@ function writeAll(fd, buffer, offset, length, position, callback) {
       } else {
         offset += written;
         length -= written;
-        position += written;
+        if (position !== null) {
```

Comment:
This is to fix 2. `position` may be `null` (see `fs.writeFile` caller) and without the check `position` will become a number instead, pointing to the middle of an existing file => corrupting data.